### DISCUSSION TITLE
Need help with animation

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#16923] [Q&A] [answered]
Hi,
I implemented a slide animation for an Expander control, but I have one issue (see code below).
(please ignore MinWidth, I needed to animate a value that I am using in binding with a converter, and Tag cannot be animated).

**The issue is that when the Expanded is loaded in the vue, the Collapse animation is played.**

In WPF, with VisualStates, I was using the parameter useTransitions=false in OnApplyTemplate, the useTransitions=true in property changed. So when the control is added to the view, the animation is not playing.

Is there a way to do the same in Avalonia? Enable/disable animation?

I tried to add another pseudoclass ":animation" and use ":expanded:animation" in a Style with an animation and ":expanded:not(:animation) in another Style with setters only, but that do not work.

May be I could use a transition, but ideally I would a different duration and easing for Expand and Collapse.

Is is a common case in WPF/WinUI, any suggestion for Avalonia?

```
<Style Selector="^:expanded /template/ Border#ExpanderContent">
    <Style.Animations>
        <Animation FillMode="Both" Duration="0:0:0.333">
            <KeyFrame Cue="0%">
                <Setter Property="MinWidth" Value="1.0" />
                <Setter Property="IsVisible" Value="True" />
            </KeyFrame>
            <KeyFrame Cue="100%" KeySpline="0.0, 0.0, 0.0, 1.0">
                <Setter Property="MinWidth" Value="0.0" />
            </KeyFrame>
        </Animation>
    </Style.Animations>
</Style>

<Style Selector="^:not(:expanded) /template/ Border#ExpanderContent">
    <Style.Animations>
        <Animation FillMode="Both" Duration="0:0:0.167">
            <KeyFrame Cue="0%">
                <Setter Property="MinWidth" Value="0.0" />
                <Setter Property="IsVisible" Value="True" />
            </KeyFrame>
            <KeyFrame Cue="100%" KeySpline="1.0, 1.0, 0.0, 1.0">
                <Setter Property="MinWidth" Value="1.0" />
                <Setter Property="IsVisible" Value="False" />
            </KeyFrame>
        </Animation>
    </Style.Animations>
</Style>
```

**Comment by maxkatz6** [comment, marked as answer]
> The issue is that when the Expanded is loaded in the vue, the Collapse animation is played.

Yes, because style selector was matched.

> I tried to add another pseudoclass ":animation" and use ":expanded:animation" in a Style with an animation and ":expanded:not(:animation) in another Style with setters only, but that do not work.

It should work. You need to add (and then remove) this class/pseudoclass on the object though.

You can also add animation to the resources (or create in code), and [run it manually](https://docs.avaloniaui.net/docs/guides/graphics-and-animation/keyframe-animations#running-animation-from-the-code-behind).

> May be I could use a transition, but ideally I would a different duration and easing for Expand and Collapse.

Transition is normally more appropriate, when you need to react on property changes. 
I think it should have been possible to set transition via styles, i.e. ":expanded" style has a setter for hiding transition, and ":not(:expanded)" the opposite - two transitions. But I never tried that, just saw it somewhere I think.

**Reply by dbriard** [reply]
Thanks for the suggestions @maxkatz6 

I initially tried the code behind path, and make it work as expect, but the code is not really clean.

Then I found what you said about Transitions in the SplitView's theme.
There is one limitation however, I have to keep the animations only for IsVisible property, because there is no BoolTransition.
I tried to make it one myself, but the code is Internal :( 
Too bad as there is everything already in place to make it one, such as the BoolAnimator
![image](https://github.com/user-attachments/assets/5b6cf2a2-8035-48d6-967d-b30857f3d561)
Or may be I missed something?

Would it be possible to add **BoolAnimation** in a future release? That would be very useful to animate panes that can Expand/Collapse with minimal xaml.

```
<Style Selector="^:expanded">
    <Style Selector="^ /template/ Border#ExpanderContent">
        <Setter Property="Transitions">
            <Transitions>
                <DoubleTransition
                    Easing="0.0, 0.0, 0.0, 1.0"
                    Property="MinWidth"
                    Duration="0:0:0.333" />
            </Transitions>
        </Setter>
        <Setter Property="MinWidth" Value="0" />
    </Style>
</Style>

<Style Selector="^:not(:expanded):down">
    <Style Selector="^ /template/ Border#ExpanderContent">
        <Setter Property="Transitions">
            <Transitions>
                <DoubleTransition
                    Easing="1.0, 1.0, 0.0, 1.0"
                    Property="MinWidth"
                    Duration="0:0:0.167" />
            </Transitions>
        </Setter>
        <Setter Property="MinWidth" Value="1" />
    </Style>
</Style>

<Style Selector="^:not(:expanded):up">
    <Style Selector="^ /template/ Border#ExpanderContent">
        <Setter Property="Transitions">
            <Transitions>
                <DoubleTransition
                    Easing="1.0, 1.0, 0.0, 1.0"
                    Property="MinWidth"
                    Duration="0:0:0.167" />
            </Transitions>
        </Setter>
        <Setter Property="MinWidth" Value="-1" />
    </Style>
</Style>

<Style Selector="^:expanded /template/ Border#ExpanderContent">
    <Style.Animations>
        <Animation FillMode="Both" Duration="0:0:0.333">
            <KeyFrame Cue="0%">
                <Setter Property="IsVisible" Value="True" />
            </KeyFrame>
        </Animation>
    </Style.Animations>
</Style>

<Style Selector="^:not(:expanded) /template/ Border#ExpanderContent">
    <Style.Animations>
        <Animation FillMode="Both" Duration="0:0:0.167">
            <KeyFrame Cue="0%">
                <Setter Property="IsVisible" Value="True" />
            </KeyFrame>
            <KeyFrame Cue="100%" KeySpline="1.0, 1.0, 0.0, 1.0">
                <Setter Property="IsVisible" Value="False" />
            </KeyFrame>
        </Animation>
    </Style.Animations>
</Style>
```

**Reply by dbriard** [reply]
@maxkatz6 Ah a last issue, I still have the jump when the view is loaded as IsVisible is set to True for 0.167 s due to the animation. 
Without BoolTransition, I think I have no other choice than implementing in code behind.
